File: classified/packages/game/cypress/support/knowledge-helpers.ts
/**
 * Knowledge Base Test Helpers
 * Utility functions for testing knowledge base CRUD operations
 */

export interface KnowledgeDocument {
  id: string;
  title: string;
  description?: string;
  createdAt: string;
  metadata?: Record<string, any>;
}

export interface KnowledgeChunk {
  id: string;
  content: {
    text: string;
    metadata?: Record<string, any>;
  };
  embedding?: number[];
}

export interface KnowledgeSearchResult {
  content: {
    text: string;
    metadata?: Record<string, any>;
  };
  similarity?: number;
}

export interface UploadResponse {
  success: boolean;
  data: {
    id: string;
    message?: string;
  };
  error?: {
    message: string;
    code?: string;
  };
}

export class KnowledgeTestHelper {
  private baseUrl = 'http://localhost:7777/knowledge';
  private agentId: string;

  constructor(agentId: string = '00000000-0000-0000-0000-000000000001') {
    this.agentId = agentId;
  }

  /**
   * Upload a file to the knowledge base
   */
  uploadFile(
    fileName: string,
    content: string,
    fileType: string = 'text/plain'
  ): Cypress.Chainable<UploadResponse> {
    const file = new File([content], fileName, { type: fileType });
    const formData = new FormData();
    formData.append('file', file);
    formData.append('agentId', this.agentId);

    return cy
      .request({
        method: 'POST',
        url: `${this.baseUrl}/documents`,
        body: formData,
        headers: {
          Accept: 'application/json',
        },
      })
      .then((response) => {
        expect(response.status).to.eq(200);
        return response.body as UploadResponse;
      });
  }

  /**
   * Upload a document from URL
   */
  uploadFromUrl(url: string): Cypress.Chainable<UploadResponse> {
    return cy
      .request({
        method: 'POST',
        url: `${this.baseUrl}/import`,
        body: {
          url,
          agentId: this.agentId,
        },
        headers: {
          'Content-Type': 'application/json',
        },
      })
      .then((response) => {
        expect(response.status).to.eq(200);
        return response.body as UploadResponse;
      });
  }

  /**
   * Get all knowledge documents
   */
  getDocuments(): Cypress.Chainable<KnowledgeDocument[]> {
    return cy
      .request({
        method: 'GET',
        url: `${this.baseUrl}/documents?agentId=${this.agentId}`,
      })
      .then((response) => {
        expect(response.status).to.eq(200);
        expect(response.body.success).to.be.true;
        return response.body.data as KnowledgeDocument[];
      });
  }

  /**
   * Get chunks for a specific document
   */
  getDocumentChunks(documentId: string): Cypress.Chainable<KnowledgeChunk[]> {
    return cy
      .request({
        method: 'GET',
        url: `${this.baseUrl}/chunks/${documentId}?agentId=${this.agentId}`,
      })
      .then((response) => {
        expect(response.status).to.eq(200);
        expect(response.body.success).to.be.true;
        return response.body.data as KnowledgeChunk[];
      });
  }

  /**
   * Search the knowledge base
   */
  search(query: string, count: number = 10): Cypress.Chainable<KnowledgeSearchResult[]> {
    return cy
      .request({
        method: 'GET',
        url: `${this.baseUrl}/search?query=${encodeURIComponent(query)}&agentId=${this.agentId}&count=${count}`,
      })
      .then((response) => {
        expect(response.status).to.eq(200);
        expect(response.body.success).to.be.true;
        return response.body.data as KnowledgeSearchResult[];
      });
  }

  /**
   * Delete a knowledge document
   */
  deleteDocument(documentId: string): Cypress.Chainable<void> {
    return cy
      .request({
        method: 'DELETE',
        url: `${this.baseUrl}/documents/${documentId}?agentId=${this.agentId}`,
      })
      .then((response) => {
        expect(response.status).to.eq(204);
      });
  }

  /**
   * Wait for document to be processed (embedded and chunked)
   */
  waitForDocumentProcessing(
    documentId: string,
    maxWaitTime: number = 30000
  ): Cypress.Chainable<void> {
    const startTime = Date.now();

    const checkProcessing = (): Cypress.Chainable<void> => {
      return this.getDocumentChunks(documentId)
        .then((chunks) => {
          if (chunks.length > 0) {
            cy.log(`Document ${documentId} processed with ${chunks.length} chunks`);
            return;
          }

          if (Date.now() - startTime > maxWaitTime) {
            throw new Error(`Document ${documentId} processing timeout after ${maxWaitTime}ms`);
          }

          cy.wait(1000);
          return checkProcessing();
        })
        .catch((error) => {
          if (Date.now() - startTime > maxWaitTime) {
            throw new Error(`Document ${documentId} processing timeout: ${error.message}`);
          }
          cy.wait(1000);
          return checkProcessing();
        });
    };

    return checkProcessing();
  }

  /**
   * Verify document exists and has content
   */
  verifyDocumentExists(documentId: string): Cypress.Chainable<boolean> {
    return this.getDocumentChunks(documentId)
      .then((chunks) => {
        return chunks.length > 0;
      })
      .catch(() => {
        return false;
      });
  }

  /**
   * Clean up - delete all test documents
   */
  cleanupTestDocuments(testFileNames: string[] = []): Cypress.Chainable<void> {
    return cy.wrap(null).then(() => {
      // Try to get documents, but continue if it fails
      return this.getDocuments()
        .then((documents) => {
          const testDocs = documents.filter(
            (doc) =>
              testFileNames.some((name) => doc.title?.includes(name)) ||
              doc.title?.includes('test-') ||
              doc.title?.includes('cypress-')
          );

          const deletePromises = testDocs.map(
            (doc) =>
              // Convert to promise for proper error handling
              new Cypress.Promise((resolve) => {
                this.deleteDocument(doc.id)
                  .then(() => resolve())
                  .catch(() => {
                    cy.log(`Failed to delete document ${doc.id}, continuing...`);
                    resolve();
                  });
              })
          );

          return cy.wrap(Promise.all(deletePromises)).then(() => {
            cy.log(`Cleaned up ${testDocs.length} test documents`);
          });
        })
        .then(
          () => {}, // Success handler
          () => {
            // Error handler for getDocuments
            cy.log('Failed to retrieve documents for cleanup, continuing...');
          }
        );
    });
  }

  /**
   * Create test documents with specific content for testing
   */
  createTestDocuments(): Cypress.Chainable<{ textDocId: string; urlDocId: string }> {
    const testTextContent =
      'This is a comprehensive test document for knowledge base testing. It contains information about artificial intelligence, machine learning, and natural language processing.';
    const testUrl = 'https://raw.githubusercontent.com/ai16z/eliza/main/README.md';

    return this.uploadFile('cypress-test-document.txt', testTextContent).then((textResponse) => {
      return this.uploadFromUrl(testUrl).then((urlResponse) => {
        return {
          textDocId: textResponse.data.id,
          urlDocId: urlResponse.data.id,
        };
      });
    });
  }

  /**
   * Validate API response structure
   */
  validateApiResponse(response: any, expectedSuccessStatus: boolean = true): void {
    expect(response).to.have.property('success', expectedSuccessStatus);

    if (expectedSuccessStatus) {
      expect(response).to.have.property('data');
    } else {
      expect(response).to.have.property('error');
      expect(response.error).to.have.property('message');
    }
  }

  /**
   * Test file upload with various file types
   */
  testFileTypes(): Cypress.Chainable<string[]> {
    const testFiles = [
      { name: 'test.txt', content: 'Plain text content', type: 'text/plain' },
      {
        name: 'test.md',
        content: '# Markdown Content\n\nThis is a test markdown file.',
        type: 'text/markdown',
      },
      { name: 'test.json', content: '{"key": "value", "test": true}', type: 'application/json' },
    ];

    const uploadPromises = testFiles.map((file) =>
      this.uploadFile(file.name, file.content, file.type).then((response) => response.data.id)
    );

    return cy.wrap(Promise.all(uploadPromises));
  }
}

// Cypress commands for easier usage in tests
declare global {
  namespace Cypress {
    interface Chainable {
      knowledgeHelper(agentId?: string): Cypress.Chainable<KnowledgeTestHelper>;
      uploadKnowledgeFile(
        fileName: string,
        content: string,
        fileType?: string
      ): Cypress.Chainable<UploadResponse>;
      searchKnowledge(query: string, count?: number): Cypress.Chainable<KnowledgeSearchResult[]>;
      deleteKnowledgeDocument(documentId: string): Cypress.Chainable<void>;
      cleanupKnowledgeTests(): Cypress.Chainable<void>;
    }
  }
}

// Register custom commands
Cypress.Commands.add('knowledgeHelper', (agentId?: string) => {
  return cy.wrap(new KnowledgeTestHelper(agentId));
});

Cypress.Commands.add(
  'uploadKnowledgeFile',
  (fileName: string, content: string, fileType?: string) => {
    const helper = new KnowledgeTestHelper();
    return helper.uploadFile(fileName, content, fileType);
  }
);

Cypress.Commands.add('searchKnowledge', (query: string, count?: number) => {
  const helper = new KnowledgeTestHelper();
  return helper.search(query, count);
});

Cypress.Commands.add('deleteKnowledgeDocument', (documentId: string) => {
  const helper = new KnowledgeTestHelper();
  return helper.deleteDocument(documentId);
});

Cypress.Commands.add('cleanupKnowledgeTests', () => {
  const helper = new KnowledgeTestHelper();
  return helper.cleanupTestDocuments(['test-', 'cypress-']);
});

export { KnowledgeTestHelper };

File: cpp_merge/classified/packages/game/cypress/support/knowledge-helpers.cpp
#include "knowledge-helpers.h"

KnowledgeTestHelper::KnowledgeTestHelper(string agentId) {
    this->agentId = agentId;
}

std::shared_ptr<Cypress::Chainable<std::shared_ptr<UploadResponse>>> KnowledgeTestHelper::uploadFile(string fileName, string content, string fileType)
{
    auto file = std::make_shared<File>(array<string>{ content }, fileName, object{
        object::pair{std:("type"), fileType}
    });
    auto formData = std::make_shared<FormData>();
    formData->append(std:("file"), file);
    formData->append(std:("agentId"), this->agentId);
    return cy->request(object{
        object::pair{std:("method"), std:("POST")}, 
        object::pair{std:("url"), string_empty + this->baseUrl + std:("/documents")}, 
        object::pair{std:("body"), formData}, 
        object::pair{std:("headers"), object{
            object::pair{std:("Accept"), std:("application/json")}
        }}
    })->then([=](auto response) mutable
    {
        expect(response["status"])->to->eq(200);
        return as<std::shared_ptr<UploadResponse>>(response["body"]);
    }
    );
}

std::shared_ptr<Cypress::Chainable<std::shared_ptr<UploadResponse>>> KnowledgeTestHelper::uploadFromUrl(string url)
{
    return cy->request(object{
        object::pair{std:("method"), std:("POST")}, 
        object::pair{std:("url"), string_empty + this->baseUrl + std:("/import")}, 
        object::pair{std:("body"), object{
            object::pair{std:("url"), std:("url")}, 
            object::pair{std:("agentId"), this->agentId}
        }}, 
        object::pair{std:("headers"), object{
            object::pair{std:("Content-Type"), std:("application/json")}
        }}
    })->then([=](auto response) mutable
    {
        expect(response["status"])->to->eq(200);
        return as<std::shared_ptr<UploadResponse>>(response["body"]);
    }
    );
}

std::shared_ptr<Cypress::Chainable<array<std::shared_ptr<KnowledgeDocument>>>> KnowledgeTestHelper::getDocuments()
{
    return cy->request(object{
        object::pair{std:("method"), std:("GET")}, 
        object::pair{std:("url"), string_empty + this->baseUrl + std:("/documents?agentId=") + this->agentId + string_empty}
    })->then([=](auto response) mutable
    {
        expect(response["status"])->to->eq(200);
        expect(response["body"]["success"])->to->be->true;
        return as<array<std::shared_ptr<KnowledgeDocument>>>(response["body"]["data"]);
    }
    );
}

std::shared_ptr<Cypress::Chainable<array<std::shared_ptr<KnowledgeChunk>>>> KnowledgeTestHelper::getDocumentChunks(string documentId)
{
    return cy->request(object{
        object::pair{std:("method"), std:("GET")}, 
        object::pair{std:("url"), string_empty + this->baseUrl + std:("/chunks/") + documentId + std:("?agentId=") + this->agentId + string_empty}
    })->then([=](auto response) mutable
    {
        expect(response["status"])->to->eq(200);
        expect(response["body"]["success"])->to->be->true;
        return as<array<std::shared_ptr<KnowledgeChunk>>>(response["body"]["data"]);
    }
    );
}

std::shared_ptr<Cypress::Chainable<array<std::shared_ptr<KnowledgeSearchResult>>>> KnowledgeTestHelper::search(string query, double count)
{
    return cy->request(object{
        object::pair{std:("method"), std:("GET")}, 
        object::pair{std:("url"), string_empty + this->baseUrl + std:("/search?query=") + encodeURIComponent(query) + std:("&agentId=") + this->agentId + std:("&count=") + count + string_empty}
    })->then([=](auto response) mutable
    {
        expect(response["status"])->to->eq(200);
        expect(response["body"]["success"])->to->be->true;
        return as<array<std::shared_ptr<KnowledgeSearchResult>>>(response["body"]["data"]);
    }
    );
}

std::shared_ptr<Cypress::Chainable<void>> KnowledgeTestHelper::deleteDocument(string documentId)
{
    return cy->request(object{
        object::pair{std:("method"), std:("DELETE")}, 
        object::pair{std:("url"), string_empty + this->baseUrl + std:("/documents/") + documentId + std:("?agentId=") + this->agentId + string_empty}
    })->then([=](auto response) mutable
    {
        expect(response["status"])->to->eq(204);
    }
    );
}

std::shared_ptr<Cypress::Chainable<void>> KnowledgeTestHelper::waitForDocumentProcessing(string documentId, double maxWaitTime)
{
    shared startTime = Date->now();
    shared checkProcessing = [=]() mutable
    {
        return this->getDocumentChunks(documentId)->then([=](auto chunks) mutable
        {
            if (chunks["length"] > 0) {
                cy->log(std:("Document ") + documentId + std:(" processed with ") + chunks["length"] + std:(" chunks"));
                return any();
            }
            if (Date->now() - startTime > maxWaitTime) {
                throw any(std::make_shared<Error>(std:("Document ") + documentId + std:(" processing timeout after ") + maxWaitTime + std:("ms")));
            }
            cy->wait(1000);
            return checkProcessing();
        }
        )->_catch([=](auto error) mutable
        {
            if (Date->now() - startTime > maxWaitTime) {
                throw any(std::make_shared<Error>(std:("Document ") + documentId + std:(" processing timeout: ") + error["message"] + string_empty));
            }
            cy->wait(1000);
            return checkProcessing();
        }
        );
    };
    return checkProcessing();
}

std::shared_ptr<Cypress::Chainable<boolean>> KnowledgeTestHelper::verifyDocumentExists(string documentId)
{
    return this->getDocumentChunks(documentId)->then([=](auto chunks) mutable
    {
        return chunks["length"] > 0;
    }
    )->_catch([=]() mutable
    {
        return false;
    }
    );
}

std::shared_ptr<Cypress::Chainable<void>> KnowledgeTestHelper::cleanupTestDocuments(array<string> testFileNames)
{
    return cy->wrap(nullptr)->then([=]() mutable
    {
        return this->getDocuments()->then([=](auto documents) mutable
        {
            shared testDocs = documents["filter"]([=](auto doc) mutable
            {
                return OR((OR((testFileNames->some([=](auto name) mutable
                {
                    return doc["title"]->includes(name);
                }
                )), (doc["title"]->includes(std:("test-"))))), (doc["title"]->includes(std:("cypress-"))));
            }
            );
            auto deletePromises = testDocs["map"]([=](auto doc) mutable
            {
                return std::make_shared<Cypress->Promise>([=](auto resolve) mutable
                {
                    this->deleteDocument(doc["id"])->then([=]() mutable
                    {
                        return resolve();
                    }
                    )->_catch([=]() mutable
                    {
                        cy->log(std:("Failed to delete document ") + doc["id"] + std:(", continuing..."));
                        resolve();
                    }
                    );
                }
                );
            }
            );
            return cy->wrap(Promise->all(deletePromises))->then([=]() mutable
            {
                cy->log(std:("Cleaned up ") + testDocs["length"] + std:(" test documents"));
            }
            );
        }
        )->then([=]() mutable
        {
        }
        , [=]() mutable
        {
            cy->log(std:("Failed to retrieve documents for cleanup, continuing..."));
        }
        );
    }
    );
}

std::shared_ptr<Cypress::Chainable<object>> KnowledgeTestHelper::createTestDocuments()
{
    auto testTextContent = std:("This is a comprehensive test document for knowledge base testing. It contains information about artificial intelligence, machine learning, and natural language processing.");
    shared testUrl = std:("https://raw.githubusercontent.com/ai16z/eliza/main/README.md");
    return this->uploadFile(std:("cypress-test-document.txt"), testTextContent)->then([=](auto textResponse) mutable
    {
        return this->uploadFromUrl(testUrl)->then([=](auto urlResponse) mutable
        {
            return object{
                object::pair{std:("textDocId"), textResponse["data"]["id"]}, 
                object::pair{std:("urlDocId"), urlResponse["data"]["id"]}
            };
        }
        );
    }
    );
}

void KnowledgeTestHelper::validateApiResponse(any response, boolean expectedSuccessStatus)
{
    expect(response)->to->have->property(std:("success"), expectedSuccessStatus);
    if (expectedSuccessStatus) {
        expect(response)->to->have->property(std:("data"));
    } else {
        expect(response)->to->have->property(std:("error"));
        expect(response["error"])->to->have->property(std:("message"));
    }
}

std::shared_ptr<Cypress::Chainable<array<string>>> KnowledgeTestHelper::testFileTypes()
{
    auto testFiles = array<object>{ object{
        object::pair{std:("name"), std:("test.txt")}, 
        object::pair{std:("content"), std:("Plain text content")}, 
        object::pair{std:("type"), std:("text/plain")}
    }, object{
        object::pair{std:("name"), std:("test.md")}, 
        object::pair{std:("content"), std:("# Markdown Content\
\
This is a test markdown file.")}, 
        object::pair{std:("type"), std:("text/markdown")}
    }, object{
        object::pair{std:("name"), std:("test.json")}, 
        object::pair{std:("content"), std:("{"key": "value", "test": true}")}, 
        object::pair{std:("type"), std:("application/json")}
    } };
    auto uploadPromises = testFiles->map([=](auto file) mutable
    {
        return this->uploadFile(file["name"], file["content"], file["type"])->then([=](auto response) mutable
        {
            return response["data"]->id;
        }
        );
    }
    );
    return cy->wrap(Promise->all(uploadPromises));
}

namespace global {
    namespace Cypress {
    }
}

void Main(void)
{
    Cypress->Commands->add(std:("knowledgeHelper"), [=](auto agentId = undefined) mutable
    {
        return cy->wrap(std::make_shared<KnowledgeTestHelper>(agentId));
    }
    );
    Cypress->Commands->add(std:("uploadKnowledgeFile"), [=](auto fileName, auto content, auto fileType = undefined) mutable
    {
        auto helper = std::make_shared<KnowledgeTestHelper>();
        return helper->uploadFile(fileName, content, fileType);
    }
    );
    Cypress->Commands->add(std:("searchKnowledge"), [=](auto query, auto count = undefined) mutable
    {
        auto helper = std::make_shared<KnowledgeTestHelper>();
        return helper->search(query, count);
    }
    );
    Cypress->Commands->add(std:("deleteKnowledgeDocument"), [=](auto documentId) mutable
    {
        auto helper = std::make_shared<KnowledgeTestHelper>();
        return helper->deleteDocument(documentId);
    }
    );
    Cypress->Commands->add(std:("cleanupKnowledgeTests"), [=]() mutable
    {
        auto helper = std::make_shared<KnowledgeTestHelper>();
        return helper->cleanupTestDocuments(array<string>{ std:("test-"), std:("cypress-") });
    }
    );
}

MAIN
